Update docs for Jetpack Gutenberg extensions

simison opened the pull request:
Elaborate more about block development in `/extensions` docs.

simison committed:
Update docs for Jetpack Gutenberg extensions

simison commented on extensions/README.md:
```diff
@@ -102,3 +102,55 @@ To test extensions for a Simple site in Calypso, sandbox the simple site URL (ex
+
+Extensions _always_ get [Gutenberg's polyfill scripts](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/scripts.md#polyfill-scripts) enqueued so you can safely use methods not supported by older browsers such as IE11.
+
+### Jetpack sidebar
```
:wave: FYI @codebykat if you wanna have a look at this section and have ideas for improvement. :-)

simison commented on extensions/README.md:
```diff
@@ -102,3 +102,55 @@ To test extensions for a Simple site in Calypso, sandbox the simple site URL (ex
+
+### Icons
+
+Please use [Material icons](https://material.io/tools/icons/?style=outline) and don't rely on icons used in core to avoid visual mixing up.
```
@scruffian @thomasguillot maybe you have more insight into this?

scruffian replied:
We've had the go-ahead to use Material icons in our projects. What do you think we need more clarity on?

simison replied:
_Why_ Material and not dashicons or other core icons:

> don't rely on icons used in core to avoid visual mixing up.

Do we use only outlined Material icons or are  filled ones ok, too?

scruffian replied:
> Why Material and not dashicons or other core icons

Because that's the Muriel guidelines

scruffian replied:
> Do we use only outlined Material icons or are filled ones ok, too?

We are going with outline

simison commented on extensions/README.md:
```diff
@@ -102,3 +102,55 @@ To test extensions for a Simple site in Calypso, sandbox the simple site URL (ex
+
+> Still confused? <a>Check out documentation for more!</a>
+
+### Colours
```
This is a messy section and doesn't give a clear indication of what to use primarily.

My understanding is that we should rely on the same colours as Gutenberg uses in the editor area unless we do something out of the ordinary when we should rely on Muriel CSS custom properties.

simison commented:
@marekhrabe @frontdevde :wave: is there anything specific you wish you would've known about Jetpack block development before starting to work on Testimonial and Logo grid blocks?

simison commented on extensions/README.md:
```diff
@@ -39,19 +39,21 @@ If your block depends on another block, place them all in extensions folder:
 1. Add your new extension's source files to the extensions/blocks directory.
-And add your extensions's slug the appropriate array in `extensions/index.json`.
+And add your extensions's slug the beta array in `extensions/index.json`.
+By keeping your extension in beta array, it's safe to do small PRs and merge frequently.
```
I've noticed block PRs tend to grow big before merged, so I'd like to emphasize this beta-aspect and small PRs for easier iteration in case it's not clear for folks.

simison commented:
From the doc:

> If your block depends on another block, place them all in extensions folder:
>
>```
>.
>├── block-name/
>└── sub-blockname/
>```

Contact info and Form block _don't_ actually follow this structure. @Automattic/poseidon do you know if there was any specific reason not to do so and if we should elaborate on docs about it so that others know what to do when implementing such blocks? Now existing examples and docs are in conflict.

scruffian commented on extensions/README.md:
```diff
@@ -39,19 +39,21 @@ If your block depends on another block, place them all in extensions folder:
 
 - Before you develop, remember to add your extension's slug to the beta array in `extensions/index.json`.
-- Ensure that, in you docker environment, the `JETPACK_BETA_BLOCKS` constant is set to `true`
+- Ensure that, in you Docker environment, the `JETPACK_BETA_BLOCKS` constant is set to `true`
```
you -> your

simison committed:
Elaborate more about sidebar

simison committed:
yours truly

roccotripaldi committed:
Update README.md

Add clarity for where to set beta block constant

roccotripaldi committed:
Update README.md

simison committed:
Add info about debugging

simison committed:
Remove "linking" section for now

I'm not sure what to write here. I assume we should document `getSiteFragment` somehow?

simison committed:
Update icons

simison committed:
Move "build" docs under "good to know" section

jeherve commented on extensions/README.md:
```diff
@@ -140,3 +130,74 @@ Since it's added to the beta array, you need to load the beta blocks as explaine
+
+> Still confused? <a>Check out documentation for more!</a>
+
+### Colours
```
We should probably stick to american spellings (e.g. color) when possible, here and downer on the page.

simison replied:
Hah, you.. can't... make ... me do i...

done in https://github.com/Automattic/jetpack/pull/12050/commits/1dd2f0aceb9f3ac8f5125261708a9cc79ce700ca

jeherve commented on extensions/README.md:
```diff
@@ -140,3 +130,74 @@ Since it's added to the beta array, you need to load the beta blocks as explaine
+
+Extensions _always_ get [Gutenberg's polyfill scripts](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/scripts.md#polyfill-scripts) enqueued so you can safely use methods not supported by older browsers such as IE11.
+
+### Jetpack sidebar
```
I don't know if "Jetpack sidebar" means much for someone who is just starting to contribute to Jetpack. Maybe we could add a screenshot here, or a little bit of explanation about the differences between blocks and plugins?

simison replied:
I left a screenshot out as those are pain to maintain but elaborated more what it is and how to find it in https://github.com/Automattic/jetpack/pull/12050/commits/22c823184c9bc4f5b44ad8c1a2ea32a229d0c656

simison committed:
Mention Likes & Shares example in sidebar docs

jeherve commented on extensions/README.md:
```diff
@@ -39,19 +39,21 @@ If your block depends on another block, place them all in extensions folder:
 ### Beta Extensions
 Generally, all new extensions should start out as beta.
 
 - Before you develop, remember to add your extension's slug to the beta array in `extensions/index.json`.
```
Should we add a mention of the WP Cli command at this point, or even higher on the page? We don't really want people to start following this step-by-step guide only to find out when they reach the last step that there was a whole paragraph about automating this downer on the page.

simison replied:
Good point. Mentioned even earlier and moved the whole section up a bit https://github.com/Automattic/jetpack/pull/12050/commits/cf46a470d7e13743bcb6b052e464a359348fc6e1

jeherve reviewed (changes requested):
Just a few quick notes as a first pass on this.

simison committed:
🇺🇸

simison committed:
Move CLI higher and mention it earlier

simison committed:
Update README.md

simison commented:
Thanks @jeherve ! I've addressed your feedback. :-)

ockham commented on extensions/README.md:
```diff
@@ -34,45 +34,89 @@ If your block depends on another block, place them all in extensions folder:
 1. Start a new branch.
 1. Add your new extension's source files to the extensions/blocks directory.
-And add your extensions's slug the appropriate array in `extensions/index.json`.
+And add your extensions' slug the beta array in `extensions/index.json`. You can use Jetpack-CLI command to scaffold the block (see below).
```
Missing `to` before `the beta array` now :grimacing:

simison commented on extensions/README.md:
````diff
@@ -140,3 +130,76 @@ Since it's added to the beta array, you need to load the beta blocks as explaine
+define( 'GUTENBERG_DEVELOPMENT_MODE', true );
+```
+
+You could add these to `docker/wordpress/wp-config.php` in your Docker environment.
````
`SCRIPT_DEBUG` won't work in `mu-plugins` since it's already defined in `wp-config.php` by default but I'll mention it here for clarity.

ockham commented on extensions/README.md:
```diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
+### Don't worry about dependencies
 
-### Files
+The build takes care of core dependencies for both editor and view -scripts. React, lodash and `@wordpress/*` [dependencies](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/scripts.md) are externalized and automatically enqueued in PHP for your extension.
```
I'd remove the `-` before `scripts` :grimacing:

ockham commented on extensions/README.md:
```diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
 
-Additionally, the slug of the new block will be added to the `beta` array in the file `extensions/index.json`.
-Since it's added to the beta array, you need to load the beta blocks as explained above to be able to test this block.
+Jetpack adds its own [plugin sidebar](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/plugin-sidebar-0/plugin-sidebar-1-up-and-running/) to the block editor. You can find it by choosing "Jetpack" from block editor's ellipsis menu or by pressing Jetpack icon in the "pinned plugins" toolbar.
```
```suggestion
Jetpack adds its own [plugin sidebar](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/plugin-sidebar-0/plugin-sidebar-1-up-and-running/) to the block editor. You can find it by choosing "Jetpack" from block the editor's ellipsis menu or by pressing the Jetpack icon in the "pinned plugins" toolbar.
```

ockham commented on extensions/README.md:
```diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
+Jetpack adds its own [plugin sidebar](https://wordpress.org/gutenberg/handbook/designers-developers/developers/tutorials/plugin-sidebar-0/plugin-sidebar-1-up-and-running/) to the block editor. You can find it by choosing "Jetpack" from block editor's ellipsis menu or by pressing Jetpack icon in the "pinned plugins" toolbar.
 
-### Examples
+The sidebar itself is always registered in the editor and filled using [Slot Fill](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/slot-fill#readme) mechanism.
```
```suggestion
The sidebar itself is always registered in the editor and filled using the [Slot Fill](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/slot-fill#readme) mechanism.
```

ockham commented on extensions/README.md:
```diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
+The sidebar itself is always registered in the editor and filled using [Slot Fill](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/slot-fill#readme) mechanism.
 
-`wp jetpack scaffold block "Cool Block"`
+Just render with `JetpackPluginSidebar` component from anywhere in your plugin's code:
```
```suggestion
Use the `JetpackPluginSidebar` component to render from anywhere in your plugin's code:
```

ockham commented on extensions/README.md:
````diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
+</JetpackPluginSidebar>
+```
+
+The sidebar won't show up at all if nothing isn't rendering in the sidebar's "slot".
````
Double negation here

```suggestion
The sidebar won't show up at all if nothing is being rendered in the sidebar's "slot".
```

ockham commented on extensions/README.md:
```diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
+
+The sidebar won't show up at all if nothing isn't rendering in the sidebar's "slot".
+
+Remember to be mindful of posts, pages, custom post types, and re-usable block post type.
```
Could be a bit clearer. Maybe

```suggestion
Remember to be mindful of what post types you want to enable your sidebar section for: e.g. posts, pages, custom post types, and re-usable block post type.
```

(Also, what's the last one?)

simison replied:
Reusable blocks are private custom post types on their own. You can find them via "manage" link in the block picker:

<img width="415" alt="image" src="https://user-images.githubusercontent.com/87168/56416318-231afb80-6299-11e9-87c0-12ae1c233597.png">

`/wp-admin/edit.php?post_type=wp_block`

I'll elaborate more in the doc.

ockham commented on extensions/README.md:
```diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
+
+Remember to be mindful of posts, pages, custom post types, and re-usable block post type.
+
+See Publicize and Shortlinks for examples how to limit functionality only to some specific post types or posts. Likes & Shares extensions are a great example of how to output content from several extensions to one sidebar section using "slots".
```
```suggestion
See [Publicize](blocks/publicize/index.js) and [Shortlinks](blocks/shortlinks/index.js) for examples how to limit functionality only to some specific post types or posts. The [Likes](blocks/likes/likes-checkbox.js) & [Sharing](blocks/sharing/sharing-checkbox.js) extensions are a great example of how to [output](shared/jetpack-likes-and-sharing-panel.js) content from several extensions to one sidebar section using "slots".
```

ockham commented on extensions/README.md:
```diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
+
+### Colors
+
+- To stay consistent with Gutenberg, your extensions should follow [Gutenberg styles and visuals]. Use Gutenberg color variables where possible.
```
Isn't `[Gutenberg styles and visuals]` missing a link?

ockham commented on extensions/README.md:
```diff
@@ -103,40 +147,66 @@ If you need to modify the build process, bear in mind that config files are also
+
+### Icons
 
+Please use outline versions of [Material icons](https://material.io/tools/icons/?style=outline) to stay in line with Muriel guidelines. Don't rely on icons used in the core to avoid visual mixing up with core blocks.
```
```suggestion
Please use outline versions of [Material icons](https://material.io/tools/icons/?style=outline) to stay in line with Muriel guidelines. Don't rely on icons used in core to avoid visual mixing up with core blocks.
```

I think we typically refer to [WordPress] Core without an article :grimacing:

ockham reviewed (dismissed):
Left a ton of nit-picks :sweat_smile: -- all of them minor. Feel free to ignore!

Good to merge :+1:

simison committed:
Mention mu-plugins for debugging constants

simison committed:
Add Gutenberg design link

simison committed:
✏️

ockham committed:
Update extensions/README.md

Co-Authored-By: simison <mikael@ihminen.org>

simison committed:
Add how to find re-usable custom post type

ockham committed:
Add links to refered blocks and files

Co-Authored-By: simison <mikael@ihminen.org>

ockham committed:
Update extensions/README.md

Co-Authored-By: simison <mikael@ihminen.org>

ockham committed:
Update extensions/README.md

Co-Authored-By: simison <mikael@ihminen.org>

ockham committed:
Update extensions/README.md

Co-Authored-By: simison <mikael@ihminen.org>

ockham committed:
Update extensions/README.md

Co-Authored-By: simison <mikael@ihminen.org>

simison commented:
@jeherve ready for Crew approval. 👍

codebykat commented:
This looks great @simison, thanks for the update 👍 

Not a blocker for this PR, but I would be interested to see more explicit docs and examples for non-block extensions such as our sidebar plugins. One of the things that tripped me up early on was that all the documentation and scaffolding was written for full-fledged blocks. e.g. the sentence "Currently the only type is block" would be a great place to say what to do if you're building something more complex.

jeherve reviewed (approved):
This is looking good. Merge when ready!

sirreal commented:
D26916-code landed ✅
